### ISSUE TITLE
Download assets in parallel

### ISSUE THREAD

**Issue by Ristovski** [mindstorm38/portablemc#16]
Right now, it seems that `portablemc` downloads all assets sequentially, which is quite slow.

A better solution would be to download all assets in parallel with a configurable amount of threads, defaulting to some sane value.

Any ideas what the official launcher does?

**Comment by mindstorm38** [comment]
I don't know how the official launcher download assets, but this is a good idea! I reserve it for the second next release.

**Comment by mindstorm38** [comment]
I was thinking about the implementation of this, in python, threads will not solve the problem because of the GIL (global interpreter lock), I will need to use sub-processes instead.

By default, the number of processes will be the number or processor cores (or processor threads?), to configure that I was thinking of an argument `--download-workers <count>` in the main command. This will be implemented in a new download function for bundles of files, the old single-file download function will be deprecated.

**Comment by Ristovski** [comment]
Perhaps `concurrent.futures` or `multiprocessing` can be used to implement a thread worker pool, but as you said, it would need to support limiting the number of threads.

Another thing is making sure the logging output does not come in out of order since that could break any progress indicators.

**Comment by mindstorm38** [comment]
Yes the logging output concerns me, this will need a little of synchronization that is enough customizable to be adapted in "richer" add-on.

**Comment by mindstorm38** [comment]
This is harder than I expected. I don't want to make this slower than before for the same amount of file. Python is not well-designed for multiprocessing, and transferring data to another processes is not simple (especially when it comes to passing strings in shared memory).

Because of that, I will begin with an improvement of the sequential download. The current sequential downloading is not really optimized because I'm opening and closing an HTTP tunnel for every file. However, I don't know how much time it will save, but this is a first step.

I had in mind a downloading system where all needed files are listed and then downloaded together in the same HTTP tunnel (or more tunnels for each CDN domain).

**Comment by Ristovski** [comment]
Something like this should work I think:

```python3
from multiprocessing.pool import ThreadPool

urls = [] # list of asset urls
NUM_THREADS = 4

def download_asset(url):
    [...]

ThreadPool(NUM_THREADS).imap_unordered(download_asset, urls)
```

**Comment by mindstorm38** [comment]
What I don't like about this solution is that it is not possible to keep a state between calls. I would prefer a system where a few processes stay running and processing downloads from a shared list or some pipe with the connection open.

**Comment by Ristovski** [comment]
Then I suppose using `requests.session()` (or some `urllib` equivalent for enabling http keep-alive) could perhaps be enough for now?

**Comment by mindstorm38** [comment]
As you can see in the branch `feature/improved-http` I'm using `http.client`, this is the lowest level I can reach, and it's really practical, so I will use that. For `requests`, it's way better than the std lib but I want to avoid external lib.

> I wouldn't be surprised if most of the current overhead is from initiating connections

Me neither! I will explore this solution!

**Comment by mindstorm38** [comment]
Main function and class are implemented! Tomorrow I will work on using this new function and testing performances.

**Comment by mindstorm38** [comment]
Full installation benchmark:
- New download system: 3min07s
- Old download system (with richer addon): 8min09s
- - Old download system (without richer addon): <missing>

**Comment by Ristovski** [comment]
Very impressive! Having it be on-par with the official launcher is amazing. Good job :)

**Comment by mindstorm38** [comment]
Pre-release sson :)

**Comment by mindstorm38** [comment]
Please test the feature or fix in the [latest pre-release](https://github.com/mindstorm38/portablemc/releases/tag/v1.1.4-pre2).
Comment this issue if the problem is fixed or not. This issue will be closed before the final release 1.1.4 if no problems are discovered.

**Comment by mindstorm38** [comment]
A real issue will be reopened specifically for the parallel download. Release 1.4 soon
